docs: How Renovate Works overview - Basic (#25108)
renovatebot/renovate: docs/usage/key-concepts/how-it-works.md

> ---

Reviewer: I think a filename like `how-renovate-works.md` is more descriptive.

> title: How it works ?

Reviewer: ```suggestion
title: How Renovate works
```

> description: Learn all about Renovate's workflow

Reviewer: ```suggestion
description: Learn how Renovate works
```

---

# Introduction

> Renovate works by searching for all the dependencies in your project and then look for updates for each.

Reviewer: ```suggestion
Renovate first finds all the dependencies in your repository, and then checks for updates to those dependencies.
```

> To be compatible with the variety of dependency-naming and versioning conventions, Renovate has defined modules for each known convention, and you can define your own too.

Reviewer: ```suggestion
Because Renovate needs to support a lot of dependency naming and versioning conventions, it has modules for each known convention.
You can define your own modules, if you want.
```

> Please add comments to the [issue#25091](https://github.com/renovatebot/renovate/issues/25091) if you wish to see a part (better) shown in the graph below

Reviewer: ```suggestion
Please comment in [Renovate issue 25091](https://github.com/renovatebot/renovate/issues/25091) if you wish to see a part (better) shown in the graph below.
```

You're closing the linked issue with this PR. After some time, _closed_ issues are automatically _locked_ by a bot:

https://github.com/renovatebot/renovate/blob/0ecc1332b3c4260dd27bc573fb08475d737fa5b2/.github/workflows/lock.yml#L19-L28

When a issue is locked, only maintainers can comment on them.

Maybe you should create a new discussion, and link to that. I think we keep discussions open forever?

https://github.com/renovatebot/renovate/blob/main/.github/workflows/lock.yml

Author: I'll create enother discussion you're right

## Modules

> The three modules are: [manager](../modules/manager/index.md), [datasource](../modules/datasource/index.md) and [versioning](../modules/versioning.md).

Reviewer: ```suggestion
Renovate's modules are:

- [manager](../modules/manager/index.md)
- [datasource](../modules/datasource/index.md)
- [versioning](../modules/versioning.md).
```

> They are used sequentially:

Reviewer: ```suggestion
Renovate uses these modules in order:
```

> 1. the manager module looks for files based on their name and extract dependencies from them, each dependency has a datasource

Reviewer: ```suggestion
1. The manager module looks for files based on their name and extracts the dependencies (each dependency has a datasource)
```

> 2. the datasource module looks for the existing versions of the dependency

Reviewer: ```suggestion
2. The datasource module looks for the existing versions of the dependency
```

> 3. the versioning module search for a valid version regarding the dependency's version

Reviewer: I don't understand this sentence. I think you mean something like:

> The versioning modules finds valid versions, based on defined rules

As far as I understand, the versioning modules has rules about what's a valid version. So the versioning module only matches known valid versions?

Author: By "valid" I meant, the one matching the rules : the higher minor, or the highest patch, of the one keeping the compatiblity suffix `-alpine`, ...

The datasource will give all existing versions, and versioning will select the good one regarding the rules

For example:

> 1. the `gitlabci` manager finds a dependency named `python:3.10-alpine` of datasource `docker`
> 2. the `docker` datasource looks for versions and finds `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`
> 3. the `docker` versioning takes `python:3.11-alpine` as it is compatible with `python:3.10-alpine`

Reviewer: ```suggestion
1. The `gitlabci` manager finds a dependency: `python:3.10-alpine` which has the  `docker` datasource
2. The `docker` datasource looks for versions and finds: `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`
3. The `docker` versioning returns `python:3.11-alpine`, because that version is compatible with `python:3.10-alpine`
```

I'm not sure how a `docker` versioning can "take a version". I think you mean returns a version?

Question for maintainers: can a versioning module _really_ know what's compatible or not? I think saying "as it is compatible, or because that version is compatible" may overstate the actual smarts in the versioning module?

# Workflow

## Basic

> Here's a high-level overview of the renovate workflow, collect dependencies then update them:

Reviewer: ```suggestion
Here's a high-level overview of Renovate's workflow, where it collects dependencies and then updates them:
```

```mermaid
flowchart LR
  subgraph COLLECT
    direction TB
    CC[[For each manager]]
    CC -->|managerA| CD["..."]
    CC -->|managerB| CCF["collect files"]

    CCF --> CFEF[[For each file]]

    CFEF -->|file1| CCD1[Collect dependency]
    CFEF -->|file2| CCD2[...]
  end

  subgraph UPDATE
    direction TB

    UC[[For each manager]]
    UC -->|managerA| UD["..."]
    UC -->|managerB| UFEF[[For each file]]

    UFEF -->|file1| FED[[For each dependency]]
    UFEF -->|file2| FED2[...]

    FED -->|dep1| D1[...]
    D1 -..-> CU
    FED -->|dep2| D2[use datasource to\n get possible updates]

>     D2 --> J[use versionning to find \n next valid update]

Reviewer: ```suggestion
    D2 --> J[use versioning to find \n next valid update]
```

    FED2 -...-> CU

    UD -....-> CU
    J --> CU[Collect updates]

    CU --> FEU[[For each update]]

    FEU --> AU[Create branch\nApply update\nSetup PR]

  end

  COLLECT --> UPDATE
```

## Advanced

> And here's a mode detailed view of the workflow:

Reviewer: ```suggestion
Here's a detailed overview of the workflow:
```

```mermaid
flowchart TB
    subgraph INITIALIZATION
        direction TB

>         MC[Merge configuration sources \n most important to least: \n cli > env > file > default]

Reviewer: I don't know what "merge configuration sources" means.

Author: It means : merging the condigurtion that comes from the cli, the env, the given config file and finally the default renovate's config

        MC --> IP[Initialize platform]
        IP --> AD[Query the platform for repositories]
        AD --> NFIL[Narrow the list with filters]
    end

    subgraph REPOSITORY
       direction TB
       FER[[For each repository]]

      subgraph COLLECTD[COLLECT DEPENDENCIES]
          direction TB
          CLBRANCH[Collect existing branches]
          CLBRANCH --> VULN[Check for vulnerabilities]
          VULN --> CC[[For each manager]]
          CC -->|managerA| CD["..."]
          CC -->|managerB| CCF["collect files"]
          CCF --> CFEF[[For each file]]
          CFEF -->|file1| CCD1[Collect dependency]
          CFEF -->|file2| CCD2[...]
      end

      subgraph COLLECTU[COLLECT UPDATES]
          direction TB
          UC[[For each manager]]
          UC -->|managerA| UD["..."]
          UC -->|managerB| UFEF[[For each file]]
          UFEF -->|file1| FED[[For each dependency]]
          UFEF -->|file2| FED2[...]
          FED -->|dep1| D1[...]
          D1 -..-> CU
          FED -->|dep2| D2[use datasource to \n get possible updates]

>           D2 --> J[use versionning to find \n next valid update]

Reviewer: ```suggestion
          D2 --> J[use versioning to find \n next valid update]
```

---
title: How it works ?
description: Learn all about Renovate's workflow
---

# Introduction

Renovate works by searching for all the dependencies in your project and then look for updates for each.

To be compatible with the variety of dependency-naming and versioning conventions, Renovate has defined modules for each known convention, and you can define your own too.

## Modules

The three modules are: [manager](../modules/manager/index.md), [datasource](../modules/datasource/index.md) and [versioning](../modules/versioning.md).

They are used sequentially:

1. the manager module looks for files based on their name and extract dependencies from them, each dependency has a datasource
2. the datasource module looks for the existing versions of the dependency
3. the versioning module search for a valid version regarding the dependency's version

For example:

> 1. the `gitlabci` manager find a dependency named `python:3.10-alpine` of datasource `docker`

Reviewer 2: ```suggestion
1. the `gitlabci` manager finds a dependency named `python:3.10-alpine` of datasource `docker`
```

> 2. the `docker` datasource looks for versions and find `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`

Reviewer 2: ```suggestion
2. the `docker` datasource looks for versions and finds `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`
```

3. the `docker` versioning takes `python:3.11-alpine` as it is compatible with `python:3.10-alpine`

```mermaid
flowchart TD
  FEM[[For each manager]]
  FEM --> FEF[[For each file]]
  FEF --> FED[[For each dependency]]
  FED --> DS[use datasources to get update list]
  DS --> VR[use versioning to find the next valid update]
  VR --> PR[Collect updates and setup the needed PRs]

> ```

Reviewer 3: missing trailing newline

---
title: How it works ?
description: Learn all about Renovate's workflow
---

# Introduction

> Renovate works by searching for all the dependencies in your projetc and then for each look for updates.

Reviewer: ```suggestion
Renovate works by searching for all the dependencies in your project and then for each look for updates.
```

> To be compatible with the variety of dependency-naming and versionning conventions, renovate has defined modules that
> holds parts of a global workflow.

Reviewer: ```suggestion
To be compatible with the variety of dependency-naming and versioning conventions, Renovate has defined modules that holds parts of a global workflow.
```

## Modules

> The three modules are: manager, datasource and versionning.

Reviewer: ```suggestion
The three modules are: manager, datasource and versioning.
```

> They are used sequentially:
> 
> 1. the manager module looks for files based on their name and extract dependencies from them, each dependency has a
>    datasource
> 2. the datasource module looks for the existing versions of the dependency
> 3. the versionning module search for a valid version regarding the dependency's version

Reviewer: ```suggestion
They are used sequentially:

1. the manager module looks for files based on their name and extract dependencies from them, each dependency has a datasource
2. the datasource module looks for the existing versions of the dependency
3. the versioning module search for a valid version regarding the dependency's version
```

> Example
> 
> 1. the `gitlabci` manager find a dependency named `python:3.10-alpine` of datasource `docker`
> 2. the `docker` datasource looks for versions and
>    find `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`
> 3. the `docker` versionning takes `python:3.11-alpine` as it is compatible with `python:3.10-alpine`

Reviewer: ```suggestion
For example:

1. the `gitlabci` manager find a dependency named `python:3.10-alpine` of datasource `docker`
2. the `docker` datasource looks for versions and find `[python:3.9,python:3.9-alpine,python:3.10,python:3.10-alpine,python:3.11,python:3.11-alpine]`
3. the `docker` versioning takes `python:3.11-alpine` as it is compatible with `python:3.10-alpine`
```

```mermaid
flowchart TD
  FEM[[For each manager]]
  FEM --> FEF[[For each file]]
  FEF --> FED[[For each dependency]]
  FED --> DS[use datasources to get update list]

>   DS --> VR[use versionning find next valid update]

Reviewer: ```suggestion
  DS --> VR[use versioning to find the next valid update]
```
